File: lib/core/crypto/encryption_service.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:encrypt/encrypt.dart' as enc;

/// AES-256-CBC end-to-end encryption service for Pacelli.
///
/// Encrypts human-readable content fields (titles, descriptions, names)
/// before they leave the device. Only household members who hold the
/// symmetric household key can decrypt.
///
/// **What is encrypted**: task titles, descriptions, subtask titles,
/// checklist titles/items, plan titles/entries/labels, category names,
/// household name, display name.
///
/// **What is NOT encrypted**: IDs, status, priority, timestamps,
/// booleans, sort orders, icons, colors — structural metadata needed
/// for Firestore queries.
class EncryptionService {
  /// Encrypts [plaintext] using AES-256-CBC with the given [key].
  /// Returns a Base64 string containing IV + ciphertext.
  ///
  /// Each call generates a fresh random IV for semantic security
  /// (identical plaintexts produce different ciphertexts).
  static String encrypt(String plaintext, String key) {
    final keyBytes = _keyFromString(key);
    final iv = enc.IV.fromSecureRandom(16);
    final encrypter = enc.Encrypter(enc.AES(keyBytes, mode: enc.AESMode.cbc));
    final encrypted = encrypter.encrypt(plaintext, iv: iv);

    // Prepend the IV so we can extract it on decrypt.
    // Format: base64(iv_bytes + ciphertext_bytes)
    final combined = Uint8List.fromList(iv.bytes + encrypted.bytes);
    return base64Encode(combined);
  }

  /// Decrypts a Base64 string produced by [encrypt] using the given [key].
  /// Returns the original plaintext.
  static String decrypt(String ciphertext, String key) {
    final keyBytes = _keyFromString(key);
    final combined = base64Decode(ciphertext);

    // First 16 bytes = IV, rest = ciphertext
    final iv = enc.IV(Uint8List.fromList(combined.sublist(0, 16)));
    final encryptedBytes = combined.sublist(16);
    final encrypted = enc.Encrypted(Uint8List.fromList(encryptedBytes));

    final encrypter = enc.Encrypter(enc.AES(keyBytes, mode: enc.AESMode.cbc));
    return encrypter.decrypt(encrypted, iv: iv);
  }

  /// Encrypts a nullable field. Returns null if input is null or empty.
  static String? encryptNullable(String? plaintext, String key) {
    if (plaintext == null || plaintext.isEmpty) return plaintext;
    return encrypt(plaintext, key);
  }

  /// Decrypts a nullable field. Returns null if input is null or empty.
  ///
  /// On decryption failure (wrong key, corrupt data, or unencrypted plaintext),
  /// logs a warning and returns `'[encrypted]'` rather than leaking ciphertext
  /// to the UI.
  static String? decryptNullable(String? ciphertext, String key) {
    if (ciphertext == null || ciphertext.isEmpty) return ciphertext;
    try {
      return decrypt(ciphertext, key);
    } catch (e) {
      // Don't leak raw ciphertext to the UI — show a safe placeholder.
      assert(() {
        // Only print in debug mode.
        // ignore: avoid_print
        print('[EncryptionService] decryptNullable failed: $e');
        return true;
      }());
      return '[encrypted]';
    }
  }

  // ── Key management helpers ──

  /// Generates a new random 256-bit key for a household.
  /// Returns the key as a 64-character hex string.
  static String generateHouseholdKey() {
    final random = Random.secure();
    final bytes = List<int>.generate(32, (_) => random.nextInt(256));
    return bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  }

  /// Legacy v1 key derivation (raw HMAC) — kept for migration only.
  static String _deriveUserKeyV1(String uid) {
    final salt = utf8.encode('pacelli_e2e_key_derivation_v1');
    final hmacSha256 = Hmac(sha256, salt);
    final digest = hmacSha256.convert(utf8.encode(uid));
    return digest.toString();
  }

  /// Derives a 256-bit user-specific key from their Firebase UID using HKDF.
  ///
  /// HKDF (HMAC-based Key Derivation Function) provides proper key stretching
  /// with extract-then-expand pattern per RFC 5869. Used only to wrap/unwrap
  /// the household key, NOT to encrypt user data directly.
  static String deriveUserKey(String uid) {
    // HKDF-Extract: create a pseudorandom key from uid + salt
    final salt = utf8.encode('pacelli_hkdf_salt_v2');
    final ikm = utf8.encode(uid);
    final prk = Hmac(sha256, salt).convert(ikm);

    // HKDF-Expand: derive output key material from PRK + info
    final info = utf8.encode('pacelli_e2e_user_key_v2');
    final expandInput = [...info, 0x01]; // info || counter byte
    final okm = Hmac(sha256, prk.bytes).convert(expandInput);

    return okm.toString(); // 64-char hex string = 256-bit key
  }

  /// Attempts to decrypt with v2 (HKDF) key first, falls back to v1 (HMAC).
  ///
  /// Used during key migration to support existing encrypted household keys.
  static String decryptKeyWithMigration(String encryptedKey, String uid) {
    // Try v2 first
    try {
      final v2Key = deriveUserKey(uid);
      return decrypt(encryptedKey, v2Key);
    } catch (_) {}

    // Fall back to v1
    final v1Key = _deriveUserKeyV1(uid);
    return decrypt(encryptedKey, v1Key);
  }

  /// Encrypts [householdKey] using [userKey] for secure storage.
  /// Used when storing the household key in Firestore per-user.
  static String encryptKeyForUser(String householdKey, String userKey) {
    return encrypt(householdKey, userKey);
  }

  /// Decrypts a stored household key using the user's derived key.
  static String decryptKeyForUser(String encryptedKey, String userKey) {
    return decrypt(encryptedKey, userKey);
  }

  // ── Internal ──

  /// Converts a 64-character hex key string to an AES Key object.
  static enc.Key _keyFromString(String hexKey) {
    final bytes = <int>[];
    for (var i = 0; i < hexKey.length; i += 2) {
      bytes.add(int.parse(hexKey.substring(i, i + 2), radix: 16));
    }
    return enc.Key(Uint8List.fromList(bytes));
  }
}

File: functions/src/crypto/encryption-service.ts
/**
 * AES-256-CBC end-to-end encryption service for Pacelli.
 *
 * TypeScript port of `lib/core/crypto/encryption_service.dart`.
 * Produces identical ciphertext format: base64(iv_16_bytes + ciphertext_bytes).
 *
 * Uses Node.js built-in `crypto` module — no external dependencies.
 */
import * as crypto from "crypto";

const ALGORITHM = "aes-256-cbc";
const IV_LENGTH = 16;
const KEY_LENGTH = 32; // 256 bits

// Key derivation constants — must match Dart exactly.
const HKDF_SALT_V2 = "pacelli_hkdf_salt_v2";
const HKDF_INFO_V2 = "pacelli_e2e_user_key_v2";
const HMAC_SALT_V1 = "pacelli_e2e_key_derivation_v1";

/**
 * Converts a 64-character hex string to a 32-byte Buffer.
 */
function keyFromHex(hexKey: string): Buffer {
  if (hexKey.length !== 64) {
    throw new Error(`Invalid key length: expected 64 hex chars, got ${hexKey.length}`);
  }
  return Buffer.from(hexKey, "hex");
}

/**
 * Encrypts plaintext using AES-256-CBC with a fresh random IV.
 * Returns base64(iv + ciphertext) — identical format to Dart.
 */
export function encrypt(plaintext: string, key: string): string {
  const keyBuf = keyFromHex(key);
  const iv = crypto.randomBytes(IV_LENGTH);

  const cipher = crypto.createCipheriv(ALGORITHM, keyBuf, iv);
  const encrypted = Buffer.concat([
    cipher.update(plaintext, "utf8"),
    cipher.final(),
  ]);

  // Prepend IV to ciphertext, then base64-encode the whole thing.
  const combined = Buffer.concat([iv, encrypted]);
  return combined.toString("base64");
}

/**
 * Decrypts a base64 string produced by `encrypt()`.
 * Splits first 16 bytes as IV, rest as ciphertext.
 */
export function decrypt(ciphertext: string, key: string): string {
  const keyBuf = keyFromHex(key);
  const combined = Buffer.from(ciphertext, "base64");

  if (combined.length < IV_LENGTH + 1) {
    throw new Error("Ciphertext too short to contain IV + data");
  }

  const iv = combined.subarray(0, IV_LENGTH);
  const encryptedBytes = combined.subarray(IV_LENGTH);

  const decipher = crypto.createDecipheriv(ALGORITHM, keyBuf, iv);
  const decrypted = Buffer.concat([
    decipher.update(encryptedBytes),
    decipher.final(),
  ]);

  return decrypted.toString("utf8");
}

/**
 * Encrypts a nullable field. Returns null/empty unchanged.
 */
export function encryptNullable(
  plaintext: string | null | undefined,
  key: string
): string | null {
  if (plaintext == null || plaintext === "") return plaintext ?? null;
  return encrypt(plaintext, key);
}

/**
 * Decrypts a nullable field. On failure returns '[encrypted]'.
 */
export function decryptNullable(
  ciphertext: string | null | undefined,
  key: string
): string | null {
  if (ciphertext == null || ciphertext === "") return ciphertext ?? null;
  try {
    return decrypt(ciphertext, key);
  } catch {
    return "[encrypted]";
  }
}

// ── Key generation ──

/**
 * Generates a new random 256-bit household key.
 * Returns a 64-character hex string — identical to Dart.
 */
export function generateHouseholdKey(): string {
  return crypto.randomBytes(KEY_LENGTH).toString("hex");
}

// ── Key derivation ──

/**
 * Legacy v1 key derivation (raw HMAC-SHA256).
 * Kept for migration only — matches Dart `_deriveUserKeyV1`.
 */
function deriveUserKeyV1(uid: string): string {
  const salt = Buffer.from(HMAC_SALT_V1, "utf8");
  const hmac = crypto.createHmac("sha256", salt);
  hmac.update(uid, "utf8");
  return hmac.digest("hex");
}

/**
 * Derives a 256-bit user key from Firebase UID using HKDF (RFC 5869).
 * Matches Dart `deriveUserKey` exactly.
 *
 * HKDF-Extract: PRK = HMAC-SHA256(salt, uid)
 * HKDF-Expand:  OKM = HMAC-SHA256(PRK, info || 0x01)
 */
export function deriveUserKey(uid: string): string {
  // HKDF-Extract
  const salt = Buffer.from(HKDF_SALT_V2, "utf8");
  const ikm = Buffer.from(uid, "utf8");
  const prk = crypto.createHmac("sha256", salt).update(ikm).digest();

  // HKDF-Expand
  const info = Buffer.from(HKDF_INFO_V2, "utf8");
  const expandInput = Buffer.concat([info, Buffer.from([0x01])]);
  const okm = crypto.createHmac("sha256", prk).update(expandInput).digest();

  return okm.toString("hex");
}

/**
 * Attempts v2 (HKDF) decryption first, falls back to v1 (HMAC).
 * Used for transparent key migration.
 */
export function decryptKeyWithMigration(encryptedKey: string, uid: string): {
  decryptedKey: string;
  wasV1: boolean;
} {
  // Try v2 first
  try {
    const v2Key = deriveUserKey(uid);
    const decryptedKey = decrypt(encryptedKey, v2Key);
    return { decryptedKey, wasV1: false };
  } catch {
    // Fall through to v1
  }

  // Fall back to v1
  const v1Key = deriveUserKeyV1(uid);
  const decryptedKey = decrypt(encryptedKey, v1Key);
  return { decryptedKey, wasV1: true };
}

/**
 * Encrypts a household key for a specific user.
 */
export function encryptKeyForUser(householdKey: string, uid: string): string {
  const userKey = deriveUserKey(uid);
  return encrypt(householdKey, userKey);
}

/**
 * Decrypts a household key using the user's derived key.
 */
export function decryptKeyForUser(encryptedKey: string, uid: string): string {
  const userKey = deriveUserKey(uid);
  return decrypt(encryptedKey, userKey);
}
